std/socket.d

```diff
@@ -307,56 +307,94 @@ bool wouldHaveBlocked() nothrow @nogc
 
-shared static this() @system
+version(Windows) private
 {
-    version(Windows)
-    {
-        WSADATA wd;
+    import core.atomic;
 
-        // Winsock will still load if an older version is present.
-        // The version is just a request.
-        int val;
-        val = WSAStartup(0x2020, &wd);
-        if (val)         // Request Winsock 2.2 for IPv6.
-            throw new SocketOSException("Unable to initialize socket library", val);
+    shared typeof(&getnameinfo) getnameinfoPointer;
+    shared typeof(&getaddrinfo) getaddrinfoPointer;
+    shared typeof(&freeaddrinfo) freeaddrinfoPointer;
 
-        // These functions may not be present on older Windows versions.
-        // See the comment in InternetAddress.toHostNameString() for details.
-        auto ws2Lib = GetModuleHandleA("ws2_32.dll");
-        if (ws2Lib)
-        {
-            getnameinfoPointer = cast(typeof(getnameinfoPointer))
-                                 GetProcAddress(ws2Lib, "getnameinfo");
-            getaddrinfoPointer = cast(typeof(getaddrinfoPointer))
-                                 GetProcAddress(ws2Lib, "getaddrinfo");
-            freeaddrinfoPointer = cast(typeof(freeaddrinfoPointer))
-                                 GetProcAddress(ws2Lib, "freeaddrinfo");
-        }
+    auto our_getnameinfo(A...)(A a) @system
+    {
+        initialize;
+        return atomicLoad(getnameinfoPointer)(a);
     }
-    else version(Posix)
+
+    auto our_getaddrinfo(A...)(A a) @system
     {
-        getnameinfoPointer = &getnameinfo;
-        getaddrinfoPointer = &getaddrinfo;
-        freeaddrinfoPointer = &freeaddrinfo;
+        initialize;
+        return atomicLoad(getaddrinfoPointer)(a);
     }
-}
 
+    auto our_freeaddrinfo(A...)(A a) @system
+    {
+        initialize;
+        return atomicLoad(freeaddrinfoPointer)(a);
+    }
 
-shared static ~this() @system nothrow @nogc
-{
-    version(Windows)
+    private void initialize() @trusted
     {
-        WSACleanup();
+        static bool tlsInitialized;
+        if (tlsInitialized) return;
+
+        static shared bool processInitialized;
+        import std.concurrency : initOnce;
+
+        static bool perform()
+        {
+            // Winsock will still load if an older version is present.
+            // The version is just a request.
+            WSADATA wd;
+            const val = WSAStartup(0x2020, &wd);
+            if (val)         // Request Winsock 2.2 for IPv6.
+                throw new SocketOSException("Unable to initialize socket library", val);
+            static extern(C) void cleanup() { WSACleanup(); }
+            atexit(&cleanup);
+
+            // These functions may not be present on older Windows versions.
+            // See the comment in InternetAddress.toHostNameString() for details.
+            auto ws2Lib = GetModuleHandleA("ws2_32.dll");
+            if (ws2Lib)
+            {
+                atomicStore(getnameinfoPointer, cast(typeof(getnameinfoPointer))
```

Comment:
If I read this correctly, it's supported since WinXP.
https://msdn.microsoft.com/en-us/library/windows/desktop/ms738520(v=vs.85).aspx#SUPPORT_FOR_GETADDRINFO_ON_WINDOWS_2000_AND_OLDER_VERSIONS_

Reply:
Yah, that's what @CyberShadow said too. Then we had linking problems so I backed off of that.

Reply:
Just seeing those linking issues myself, it's an outed ws2_32.lib (OMF import lib for optlink) shipping with dmd, so much to that rotten format.

Reply:
> Yah, that's what @CyberShadow said too. Then we had linking problems so I backed off of that.

I just fixed those linking issues, https://github.com/dlang/installer/pull/264, after 6 years :o, but it'll take a while till those libs arrive at the testers.